### PR TITLE
qa/tests: add ceph-deploy upgrade tests

### PR DESCRIPTION
**Pull request by vasukulkarni**
Add upgrade test using ceph-deploy which closely mimics admin like scenario.

### CODE REVIEW

**Comment by vasukulkarni** [comment]
Depends on https://github.com/ceph/ceph-deploy/pull/448

**Comment by vasukulkarni** [comment]
Waiting on resolution on http://tracker.ceph.com/issues/20950 to fully test.  The yaml meta has full description of the test.  Would like to backport to luminous as well.

**Comment by vasukulkarni** [comment]
Recent run: http://pulpito.ceph.com/vasu-2017-08-08_18:34:23-upgrade-master-distro-basic-vps/

File: qa/tasks/ceph_deploy.py [hunk 1]
```diff
@@ -606,11 +618,11 @@ def execute_cdeploy(admin, cmd, path):
     log.info("Waiting for cluster to become healthy")
     with contextutil.safe_while(sleep=10, tries=6,
                                 action='check health') as proceed:
-       while proceed():
-           r = remote.run(args=['sudo', 'ceph', 'health'], stdout=StringIO())
-           out = r.stdout.getvalue()
-           if (out.split(None,1)[0] == 'HEALTH_OK'):
-               break
+        while proceed():
```
Comment (amitkumar50) [inline qa/tasks/ceph_deploy.py]: I believe extra spaces are left?

File: qa/tasks/ceph_deploy.py [hunk 2]
```diff
@@ -678,6 +690,93 @@ def single_node_test(ctx, config):
...
+            log.info("Upgrading ceph on  %s", nodename)
+            ceph_admin.run(
+                args=[
+                    'cd',
+                    '{tdir}/ceph-deploy'.format(tdir=testdir),
+                    run.Raw('&&'),
+                    run.Raw(cmd),
+                ],
+            )
+            # restart all ceph services, ideally upgrade should but it doesn't
```
Comment (amitkumar50) [inline qa/tasks/ceph_deploy.py]: Please use "Does not" instead of doesn't as per coding guidelines.
A PR is raised/accepted for the same.

**Comment by vasukulkarni** [comment]
Finally with workaround ( commit 607d22cb128dd37c9849b781d6161d1e51266ed1 )to recreate mgr key, the jewel -> luminous works but kraken -> luminous seems to be a genuine issue (will look and file an issue) 

http://pulpito.ceph.com/vasu-2017-08-24_20:59:02-upgrade-master-distro-basic-vps/

**Comment by liewegas** [comment]
needs rebase

File: qa/suites/upgrade/jewel-x/ceph-deploy/jewel-luminous.yaml [hunk 3]
```diff
@@ -0,0 +1,61 @@
+meta:
+- desc: |
+    Setup 4 node ceph cluster using ceph-deploy, use latest
+    stable jewel as initial release, upgrade to luminous and
+    also setup mgr nodes along after upgrade, check for
+    cluster to reach healthy state, After upgrade run kernel tar/untar
```
Comment (amitkumar50) [inline qa/suites/upgrade/jewel-x/ceph-deploy/jewel-luminous.yaml]: may be to use full stop after `healthy state`

Comment (vasukulkarni) [reply qa/suites/upgrade/jewel-x/ceph-deploy/jewel-luminous.yaml]: I feel comma is still appropriate there since its explaining the full test.

**Comment by vasukulkarni** [comment]
@liewegas rebased done, recent run: http://pulpito.ceph.com/vasu-2017-08-31_15:50:09-upgrade-master-distro-basic-vps/ ,  I have filed an issue for kraken -> luminous.  This is mainly for luminous backport. For master, I will update the yaml to go from jewel(stable) -> luminous(stable) -> master(dev) and for luminous branch it will be just jewel(stable) -> luminous(dev).

File: qa/tasks/ceph_deploy.py [hunk 4]
```diff
@@ -145,7 +147,7 @@ def get_nodes_using_role(ctx, target_role):
 
     # Prepare a modified version of cluster.remotes with ceph-deploy-ized names
     modified_remotes = {}
-
+    ceph_deploy_mapped = dict()
```
Comment (amitkumar50) [inline qa/tasks/ceph_deploy.py]: tab left

**Comment by vasukulkarni** [comment]
retest this please

**Comment by tmuthamizhan** [comment]
lgtm

**Comment by tmuthamizhan** [comment]
for the record : http://tracker.ceph.com/issues/21122 is filed for the failing upgrade run Kraken to luminous.
run:  http://pulpito.ceph.com/vasu-2017-08-31_15:50:09-upgrade-master-distro-basic-vps/

**Pull request merged**